src/tiledb/cloud/utilities/consolidate.py

```diff
@@ -0,0 +1,248 @@
+from collections import defaultdict
+from os.path import basename
+from typing import Any, Mapping, Optional, Sequence
+
+import tiledb
+from tiledb.cloud import dag
+from tiledb.cloud.rest_api.models import RetryStrategy
+from tiledb.cloud.utilities import get_logger
+from tiledb.cloud.utilities import max_memory_usage
+from tiledb.cloud.utilities import run_dag
+
+MAX_FRAGMENT_SIZE = 1 << 30  # bytes
+
+
+def group_fragments(
+    array_uri: str,
+    *,
+    config: Optional[Mapping[str, Any]] = None,
+    group_first_dim: bool = True,
+) -> Sequence[Sequence[tiledb.FragmentInfo]]:
+    """
+    Get a list of fragment info objects, optionally grouping fragments that have the
+    same value for the first dimension.
+
+    :param array_uri: array URI
+    :param config: config dictionary, defaults to None
+    :param group_first_dim: group by first dimension, defaults to True
+    :return: list of lists of fragment info objects
+    """
+
+    logger = get_logger()
+    logger.info("Grouping fragments for array %r", array_uri)
+
+    with tiledb.scope_ctx(config):
+        fragment_info_lists = defaultdict(list)
+
+        fis = tiledb.FragmentInfoList(array_uri)
+
+        for fi in fis:
+            if group_first_dim:
+                fragment_info_lists[fi.nonempty_domain[0]].append(fi)
+            else:
+                fragment_info_lists["all"].append(fi)
+
+        # Create a list of lists of fragment info objects.
+        results = [x for x in list(fragment_info_lists.values())]
+
+        logger.info(f"{len(fis)} fragments grouped into {len(results)} groups")
+        logger.info(f"max memory usage: {max_memory_usage() / (1 << 30):.3f} GiB")
+
+        return results
+
+
+def consolidate(
+    array_uri: str,
+    fragments: Sequence[tiledb.FragmentInfo],
+    *,
+    config: Optional[Mapping[str, Any]] = None,
+    max_fragment_size: int = MAX_FRAGMENT_SIZE,
+) -> None:
+    """
+    Consolidate fragments
+
+    :param array_uri: array URI
+    :param fragments: list of fragments
+    :param config: config dictionary, defaults to None
+    :param max_fragment_size: max size of consolidated fragments,
+        defaults to MAX_FRAGMENT_SIZE
+    """
+
+    logger = get_logger()
+    logger.info(f"Consolidating {len(fragments)} fragments")
+
+    config = tiledb.Config(config)
+    config["sm.consolidation.mode"] = "fragments"
+    config["sm.consolidation.max_fragment_size"] = max_fragment_size
+
+    # Consolidate fragments.
+    fragment_names = [basename(fi.uri) for fi in fragments]
+    with tiledb.open(array_uri, "w", config=config) as array:
+        array.consolidate(fragment_uris=fragment_names)
+
+    logger.info(f"max memory usage: {max_memory_usage() / (1 << 30):.3f} GiB")
+
+
+def convac(
```

Comment:
I’m going back and forth on whether I like this name. On the one hand, it is snappy; on the other, it is maybe a bit *too* short, and could be a bit opaque until you already know what it does. That said, it is described in the first three words of the docstring. Not a hard request, just something to think about. If you want to keep it, it’s absolutely fine; I leave it up to your judgement.

Reply:
I feel the same, the name is a little too short and not descriptive. I'll pick a different name.